### PR TITLE
Feature/auth

### PR DESCRIPTION
**Pull request by mcelaney**

### CODE REVIEW

File: app/assets/javascripts/sessions.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (mcelaney) [inline app/assets/javascripts/sessions.coffee]: Remove auto-generated file

File: app/assets/stylesheets/sessions.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the sessions controller here.
```
Comment (mcelaney) [inline app/assets/stylesheets/sessions.scss]: remove auto-generated file

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  # before_filter :authorize
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    user = User.find_by_email(params[:email])
+    # if user exists with entered password is correct
+    if user && user.authenticate(params[params[:password]])
```
Comment (mcelaney) [inline app/controllers/sessions_controller.rb]: Move user to a method

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      # save the user id
+      session[:user_id] = user.id
+      redirect_to root_url, flash[:notice] = 'Logged in!'
```
Comment (mcelaney) [inline app/controllers/sessions_controller.rb]: flash is a hash... 

redirect_to root_url, notice: 'Logged in!'

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#
+class User < ActiveRecord::Base
+  has_secure_password
+
+  validates :email,
+            presence: true,
+            uniqueness: true,
```
Comment (mcelaney) [inline app/models/user.rb]: This is not case sensitive...

brian.e.mcelaney@gmail.com and Brian.E.McElaney@gmail.com

would be able to register for two seperate accounts

File: app/views/sessions/new.html.haml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+%h1 Login
+
+=form_for(@user, url: login_url) do |f|
+
+  .field
+  =f.label :email
```
Comment (mcelaney) [inline app/views/sessions/new.html.haml]: indent to fall under element

File: app/views/sessions/new.html.haml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+%h1 Login
+
+=form_for(@user, url: login_url) do |f|
```
Comment (mcelaney) [inline app/views/sessions/new.html.haml]: Change form like we talked about

File: spec/controllers/sessions_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
```
Comment (mcelaney) [inline spec/controllers/sessions_controller_spec.rb]: Delete controller spec

File: spec/features/auth_spec.rb [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      expect(page).to have_link('Login')
+      expect(page).to_not have_content('Logout')
+      expect(page).to have content('invalid')
+      expect(page).to have content('Sign Up')
+    end
+  end
+
+  context 'As a credentialed user' do
+    scenario 'I can log out' do
+      visit root_url
```
Comment (mcelaney) [inline spec/features/auth_spec.rb]: Need to log in first (create_current_user helper)

File: spec/features/auth_spec.rb [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+             email: email,
+             password: password,
+             password_confirmation: password
+            )
+    end
+
+    context 'providing valid login information' do
+      scenario 'I can authenticate as a user of collections' do
+        visit root_url
+        expect(page).to have_content('Login')
```
Comment (mcelaney) [inline spec/features/auth_spec.rb]: redundant

File: spec/features/auth_spec.rb [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  context 'providing invalid login information' do
+    let(:email) { 'This email is invalid' }
+    let(:password) { 'This password is an invalid password' }
+    let(:password_confirmation) { 'This password is an invalid password' }
+
+    scenario 'I can see errors' do
+      visit root_url
+      expect(page).to have_content('Login')
```
Comment (mcelaney) [inline spec/features/auth_spec.rb]: redundant

File: spec/features/auth_spec.rb [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+      expect(page).to have content('Sign Up')
+    end
+  end
+
+  context 'As a credentialed user' do
+    scenario 'I can log out' do
+      visit root_url
+      expect(page).to have_content('You have logged out')
+      expect(page).to_not have('Logout')
+      expect(page).to have_link
```
Comment (mcelaney) [inline spec/features/auth_spec.rb]: have_link ?

File: spec/helpers/sessions_helper_spec.rb [hunk 13]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
+
```
Comment (mcelaney) [inline spec/helpers/sessions_helper_spec.rb]: remove generated file (unless you end up with a session helper)

File: spec/views/sessions/new.html.haml_spec.rb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe 'sessions/new.html.haml', type: :view do
```
Comment (mcelaney) [inline spec/views/sessions/new.html.haml_spec.rb]: no view specs

File: app/assets/javascripts/sessions.coffee (hunk 1 above)
Comment (Dana-G) [inline app/assets/javascripts/sessions.coffee]: done

File: app/assets/stylesheets/sessions.scss (hunk 2 above)
Comment (Dana-G) [inline app/assets/stylesheets/sessions.scss]: done

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (Dana-G) [inline app/controllers/sessions_controller.rb]: think is done, also params corrected

File: app/controllers/sessions_controller.rb (hunk 4 above)
Comment (Dana-G) [inline app/controllers/sessions_controller.rb]: done

File: app/models/user.rb (hunk 5 above)
Comment (Dana-G) [inline app/models/user.rb]: replaced uniqueness: { case_sensitive: true }

File: app/views/sessions/new.html.haml (hunk 6 above)
Comment (Dana-G) [inline app/views/sessions/new.html.haml]: done

File: app/views/sessions/new.html.haml (hunk 7 above)
Comment (Dana-G) [inline app/views/sessions/new.html.haml]: bah! done, i think

File: spec/controllers/sessions_controller_spec.rb (hunk 8 above)
Comment (Dana-G) [inline spec/controllers/sessions_controller_spec.rb]: byebye'd

File: spec/features/auth_spec.rb (hunk 9 above)
Comment (Dana-G) [inline spec/features/auth_spec.rb]: done

File: spec/features/auth_spec.rb (hunk 10 above)
Comment (Dana-G) [inline spec/features/auth_spec.rb]: no more

File: spec/features/auth_spec.rb (hunk 11 above)
Comment (Dana-G) [inline spec/features/auth_spec.rb]: done

File: spec/features/auth_spec.rb (hunk 12 above)
Comment (Dana-G) [inline spec/features/auth_spec.rb]: no idea. deleted for now

File: spec/helpers/sessions_helper_spec.rb (hunk 13 above)
Comment (Dana-G) [inline spec/helpers/sessions_helper_spec.rb]: seems not, so far. deleted. Also deleted the non-spec session helper file. Will put back if need it.

File: spec/views/sessions/new.html.haml_spec.rb (hunk 14 above)
Comment (Dana-G) [inline spec/views/sessions/new.html.haml_spec.rb]: view spec file and related spec>view>session folder deleted

**Pull request merged**